### ISSUE TITLE
Add labels and annotations to container to support debugging and operational tasks

### ISSUE THREAD

**Issue by ptitzler** [elyra-ai/elyra#1277]
When a pipeline node is executed, KFP launches a pod, which executes the node's Python script or notebook. The pod name, labels, and annotations are auto-generated and don't contain any information that associates the pod with the node. Therefore currently the only way to associate a node with a pod is to inspect the KFP UI and/or? CLI , which doesn't allow for easy reverse lookup ("determine which pipeline and node is associated with pod XYZ that is [insert something that has a negative impact on the system]") 

![image](https://user-images.githubusercontent.com/13068832/107086034-3bca0d80-67ae-11eb-8b43-1e364843ff66.png)

Example list of pods in one of our Kubernetes test environments:

```
$ kubectl get pods --namespace anonymous
NAME                                              READY   STATUS      RESTARTS   AGE
aaa-c5x7s-1198940926                              0/2     Completed   0          20h
lambda-2rvgx-3787107965                           0/2     Completed   0          21h
lambda-4wbg9-3142036415                           0/2     Completed   0          5h38m
lambda-qz7l8-3348423942                           0/2     Completed   0          23h
lambda-qz7l8-3574047449                           0/2     Completed   0          23h
lambda-s4cfs-777187791                            0/2     Completed   0          20h
ml-pipeline-ui-artifact-c8fbcb8f9-km6x6           2/2     Running     0          2d21h
ml-pipeline-visualizationserver-6b78c9646-p6lkm   2/2     Running     0          2d21h
untitled-48cqr-1098260914                         0/2     Completed   0          3h9m
```
Inspecting `lambda-2rvgx-3787107965 `, which embodies a pipeline node

```
$ kubectl describe pod lambda-2rvgx-3787107965 --namespace anonymous
Name:         lambda-2rvgx-3787107965
Namespace:    anonymous
Priority:     0
Node:         devises1.fyre.ibm.com/9.30.211.16
Start Time:   Thu, 04 Feb 2021 14:41:06 -0800
Labels:       pipelines.kubeflow.org/cache_enabled=true
              pipelines.kubeflow.org/cache_id=
              pipelines.kubeflow.org/metadata_context_id=7
              pipelines.kubeflow.org/metadata_execution_id=12
              pipelines.kubeflow.org/metadata_written=true
              workflows.argoproj.io/completed=true
              workflows.argoproj.io/workflow=lambda-2rvgx
Annotations:  pipelines.kubeflow.org/execution_cache_key: 8f6d7dfa82faa95956d0b182db12021712e01161a06711cbfef332a755017b0e
              pipelines.kubeflow.org/metadata_input_artifact_ids: []
              pipelines.kubeflow.org/metadata_output_artifact_ids:
                [{"id": 11, "name": "mlpipeline-ui-metadata", "uri": "minio://mlpipeline/artifacts/lambda-2rvgx/lambda-2rvgx-3787107965/mlpipeline-ui-meta...
              sidecar.istio.io/inject: false
              workflows.argoproj.io/node-name: lambda-2rvgx.untitled
              workflows.argoproj.io/outputs:
                {"artifacts":[{"name":"mlpipeline-metrics","path":"/tmp/mlpipeline-metrics.json","optional":true},{"name":"mlpipeline-ui-metadata","path":...
              workflows.argoproj.io/template:
                {"name":"untitled","inputs":{},"outputs":{"artifacts":[{"name":"mlpipeline-metrics","path":"/tmp/mlpipeline-metrics.json","optional":true}...
Status:       Succeeded
IP:           172.17.0.61
IPs:
  IP:           172.17.0.61
Controlled By:  Workflow/lambda-2rvgx
Containers:
  wait:
    Container ID:  docker://51edfc93aff09c08fc2056a456c75a112bef17057c736bf00bf88a2d8c3af51c
    Image:         argoproj/argoexec:v2.3.0
    Image ID:      docker-pullable://argoproj/argoexec@sha256:85132fc2c8bc373fca885df17637d5d35682a23de8d1390668a5e1c149f2f187
    Port:          <none>
    Host Port:     <none>
    Command:
      argoexec
      wait
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Thu, 04 Feb 2021 14:41:09 -0800
      Finished:     Thu, 04 Feb 2021 14:41:44 -0800
    Ready:          False
    Restart Count:  0
    Environment:
      ARGO_POD_NAME:  lambda-2rvgx-3787107965 (v1:metadata.name)
    Mounts:
      /argo/podmetadata from podmetadata (rw)
      /argo/secret/mlpipeline-minio-artifact from mlpipeline-minio-artifact (ro)
      /var/run/docker.sock from docker-sock (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from default-editor-token-6j9xd (ro)
  main:
    Container ID:  docker://9253f7abda499c6b59d86f74870134c21490069e4aa0882800dac16f8644a83a
    Image:         quay.io/ptitzler/pandas:1.1.4
    Image ID:      docker-pullable://quay.io/ptitzler/pandas@sha256:b55a9e1a15250aa86f5fd6ddb6025a12b295dae6e919f31479be269622d9d78c
    Port:          <none>
    Host Port:     <none>
    Command:
      sh
      -c
    Args:
      mkdir -p ./jupyter-work-dir/ && cd ./jupyter-work-dir/ && curl -H "Cache-Control: no-cache" -L https://raw.githubusercontent.com/ptitzler/kfp-notebook/label/etc/docker-scripts/bootstrapper.py --output bootstrapper.py && curl -H "Cache-Control: no-cache" -L https://raw.githubusercontent.com/ptitzler/kfp-notebook/label/etc/requirements-elyra.txt --output requirements-elyra.txt && python3 -m pip install  packaging && python3 -m pip freeze > requirements-current.txt && python3 bootstrapper.py --cos-endpoint http://devises1.fyre.ibm.com:30927 --cos-bucket pipeline-artifacts --cos-directory "untitled-0204144104" --cos-dependencies-archive "Untitled-eae61c44-e3df-46c8-ac41-60a2c6e93c7a.tar.gz" --file "Untitled.ipynb" 
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Thu, 04 Feb 2021 14:41:10 -0800
      Finished:     Thu, 04 Feb 2021 14:41:43 -0800
    Ready:          False
    Restart Count:  0
    Environment:
      AWS_ACCESS_KEY_ID:             REDACTED
      AWS_SECRET_ACCESS_KEY:         REDACTED
      ELYRA_ENABLE_PIPELINE_INFO:    True
      ELYRA_WRITABLE_CONTAINER_DIR:  /tmp
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-editor-token-6j9xd (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  podmetadata:
    Type:  DownwardAPI (a volume populated by information about the pod)
    Items:
      metadata.annotations -> annotations
  docker-sock:
    Type:          HostPath (bare host directory volume)
    Path:          /var/run/docker.sock
    HostPathType:  Socket
  mlpipeline-minio-artifact:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  mlpipeline-minio-artifact
    Optional:    false
  default-editor-token-6j9xd:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-editor-token-6j9xd
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:          <none>
```

there appears to be not much identifying information.
By adding labels and annotations, that for example identify in a unique way the pipeline and the node the problem could be resolved. With labels, for example, one could easily query the system to only return pods that belong to a specific pipeline.

**Comment by ptitzler** [comment]
An implementation would require delivery of https://github.com/elyra-ai/elyra/pull/1276, which implicitly makes new APIs available that allow for addition of labels and annotations to container ops.

**Comment by ptitzler** [comment]
Background information for Kubernetes [labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/) and [annotations](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/).

TL;DR
> You can use either labels or annotations to attach metadata to Kubernetes objects. Labels can be used to select objects and to find collections of objects that satisfy certain conditions. In contrast, annotations are not used to identify and select objects. The metadata in an annotation can be small or large, structured or unstructured, and can include characters not permitted by labels.

**Comment by ptitzler** [comment]
Metadata we should attach to each pod that processes a node:
 - Elyra pipeline name (the pipeline file name without extension)
 - Elyra pipeline run name (as provided by the user)
 - Elyra pipeline node name (which is displayed in the Elyra UI in the editor)
 - Elyra node type


Need to take into account that
> Labels are key/value pairs. Valid label keys have two segments: an optional prefix and name, separated by a slash (/). The name segment is required and must be 63 characters or less, beginning and ending with an alphanumeric character ([a-z0-9A-Z]) with dashes (-), underscores (_), dots (.), and alphanumerics between. The prefix is optional. If specified, the prefix must be a DNS subdomain: a series of DNS labels separated by dots (.), not longer than 253 characters in total, followed by a slash (/).

> If the prefix is omitted, the label Key is presumed to be private to the user. Automated system components (e.g. kube-scheduler, kube-controller-manager, kube-apiserver, kubectl, or other third-party automation) which add labels to end-user objects must specify a prefix.

> Valid label values must be 63 characters or less and must be empty or begin and end with an alphanumeric character ([a-z0-9A-Z]) with dashes (-), underscores (_), dots (.), and alphanumerics between.

and
> Annotations are key/value pairs. Valid annotation keys have two segments: an optional prefix and name, separated by a slash (/). The name segment is required and must be 63 characters or less, beginning and ending with an alphanumeric character ([a-z0-9A-Z]) with dashes (-), underscores (_), dots (.), and alphanumerics between. The prefix is optional. If specified, the prefix must be a DNS subdomain: a series of DNS labels separated by dots (.), not longer than 253 characters in total, followed by a slash (/).
